contracts/upgradeable_contracts/BridgeValidators.sol

```diff
@@ -9,89 +9,167 @@ import "../upgradeability/EternalStorage.sol";
+            } else {
+                setValidator(_initialValidators[i - 1], _initialValidators[i]);
+            }
+
             setValidatorCount(validatorCount().add(1));
-            setValidator(_initialValidators[i], true);
+            setValidatorRewardAddress(_initialValidators[i], _initialRewards[i]);
             emit ValidatorAdded(_initialValidators[i]);
         }
+
         uintStorage[keccak256(abi.encodePacked("requiredSignatures"))] = _requiredSignatures;
         uintStorage[keccak256("deployedAtBlock")] = block.number;
         setInitialize(true);
         emit RequiredSignaturesChanged(_requiredSignatures);
+
         return isInitialized();
     }
 
-    function addValidator(address _validator) external onlyOwner {
-        require(_validator != address(0));
+    function addValidator(address _validator, address _reward) external onlyOwner {
+        require(_validator != address(0) && _validator != F_ADDR);
+        require(_reward != address(0));
         require(!isValidator(_validator));
+
+        address firstValidator = getNextValidator(F_ADDR);
+        setValidator(_validator, firstValidator);
+        setValidatorRewardAddress(_validator, _reward);
+        setValidator(F_ADDR, _validator);
+
         setValidatorCount(validatorCount().add(1));
-        setValidator(_validator, true);
         emit ValidatorAdded(_validator);
     }
 
     function removeValidator(address _validator) external onlyOwner {
         require(validatorCount() > requiredSignatures());
         require(isValidator(_validator));
-        setValidator(_validator, false);
+        address validatorsNext = getNextValidator(_validator);
+        address index = F_ADDR;
+        address next = getNextValidator(index);
+
+        // find the element in the list pointing to _validator
+        while (next != _validator) {
+            index = next;
+            next = getNextValidator(index);
+        }
+
+        setValidator(index, validatorsNext);
+        deleteItemFromAddressStorage("validatorsList", _validator);
+        deleteItemFromAddressStorage("validatorsRewards", _validator);
         setValidatorCount(validatorCount().sub(1));
+
         emit ValidatorRemoved(_validator);
     }
 
-    function setRequiredSignatures(uint256 _requiredSignatures) external onlyOwner {
+    function setRequiredSignatures(uint256 _requiredSignatures)
+        external
+        onlyOwner
+    {
         require(validatorCount() >= _requiredSignatures);
         require(_requiredSignatures != 0);
         uintStorage[keccak256(abi.encodePacked("requiredSignatures"))] = _requiredSignatures;
         emit RequiredSignaturesChanged(_requiredSignatures);
     }
 
-    function getBridgeValidatorsInterfacesVersion() public pure returns(uint64 major, uint64 minor, uint64 patch) {
+    function getBridgeValidatorsInterfacesVersion()
+        public
+        pure
+        returns (uint64 major, uint64 minor, uint64 patch)
+    {
         return (2, 0, 0);
```

Comment:
We need to increase the version of the contract.

Reply:
Done on https://github.com/poanetwork/poa-bridge-contracts/pull/130/commits/349bbbb355d9ee155eefc53886780a4e6f3f7f4c